Shared/Tests.Shared/InstanceTests.cs

```diff
@@ -227,5 +228,134 @@ public void RealmObjectClassesOnlyAllowRealmObjects()
         }
+
+        [TestCase(true, true)]
+        [TestCase(true, false)]
+        [TestCase(false, true)]
+        [TestCase(false, false)]
+        public void Compact_ShouldReduceSize(bool encrypt, bool populate)
+        {
+            var config = new RealmConfiguration($"compactrealm_{encrypt}_{populate}.realm");
+            if (encrypt)
+            {
+                config.EncryptionKey = new byte[64];
+                config.EncryptionKey[0] = 5;
+            }
+
+            using (var realm = Realm.GetInstance(config))
+            {
+                if (populate)
+                {
+                    AddDummyData(realm);
+                }
+            }
+
+            var initialSize = new FileInfo(config.DatabasePath).Length;
+
+            Assert.That(Realm.Compact(config));
+
+            var finalSize = new FileInfo(config.DatabasePath).Length;
+            Assert.That(initialSize >= finalSize);
+
+            using (var realm = Realm.GetInstance(config))
+            {
+                Assert.That(realm.All<IntPrimaryKeyObject>().Count(), Is.EqualTo(populate ? 500 : 0));
+            }
+        }
+
+        [Test]
+        public void Compact_WhenInTransaction_ShouldThrow()
+        {
+            using (var realm = Realm.GetInstance())
+            {
+                Assert.That(() =>
+                {
+                    realm.Write(() =>
+                    {
+                        Realm.Compact();
+                    });
+                }, Throws.TypeOf<RealmInvalidTransactionException>());
+            }
+        }
+
+        [Test]
+        public void Compact_WhenOpenOnDifferentThread_ShouldReturnFalse()
+        {
+            using (var realm = Realm.GetInstance())
+            {
+                AddDummyData(realm);
+
+                var initialSize = new FileInfo(realm.Config.DatabasePath).Length;
+                Assert.IsFalse(Task.Run(() => Realm.Compact(realm.Config)).Result);
+                var finalSize = new FileInfo(realm.Config.DatabasePath).Length;
+
+                Assert.That(finalSize, Is.EqualTo(initialSize));
+            }
+        }
+
+        [Test, Ignore("Currently doesn't work. Ref #947")]
+        public void Compact_WhenOpenOnSameThread_ShouldReturnFalse()
+        {
+            // TODO: enable when we implement instance caching (#947)
+            // This works because of caching of native instances in ObjectStore.
+            // Technically, we get the same native instance, so Compact goes through.
+            // However, this invalidates the opened realm, but we have no way of communicating that.
+            // That is why, things seem fine until we try to run queries on the opened realm.
+            // Once we handle caching in managed, we should reenable the test.
+            using (var realm = Realm.GetInstance())
+            {
+                var initialSize = new FileInfo(realm.Config.DatabasePath).Length;
+                Assert.IsFalse(Realm.Compact());
```

Comment:
I prefer `Assert.That(.., Is.False)` but for whatever reason I use `Assert.Throws` so I should get off my high horse...